pkg/plugins/manager/pipeline/discovery/discovery.go

```diff
@@ -0,0 +1,64 @@
+package discovery
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/config"
+	"github.com/grafana/grafana/pkg/plugins/log"
+)
+
+// Discoverer is responsible for the Discovery stage of the plugin loader pipeline.
+type Discoverer interface {
+	Discover(ctx context.Context, src plugins.PluginSource) ([]*plugins.FoundBundle, error)
+}
+
+// FindFunc is the function used for the Find step of the Discovery stage.
+type FindFunc func(ctx context.Context, src plugins.PluginSource) ([]*plugins.FoundBundle, error)
+
+// FindFilterFunc is the function used for the Filter step of the Discovery stage.
+type FindFilterFunc func(ctx context.Context, class plugins.Class, bundles []*plugins.FoundBundle) ([]*plugins.FoundBundle, error)
```

Comment:
I am a bit confused by the difference between the Find-level filter and the verification step.

Shouldn't the verification step take care of filtering out plugins? Like for plugins with invalid signatures, or angular plugins (when they are disabled?)

> Verification: Verify the plugins based on some criteria (e.g. signature validation, angular detection, etc.)

Reply:
I don't have any strong feelings on this yet without having tackled it properly. There's definitely some blurry lines and I think as we progress, a lot of this stuff will get moved around into different places, so you could absolutely be right there. Maybe we tackle that one next and see what makes the most sense.

My initial thought was that the verification stage felt like a verification of the plugin itself, whereas the duplication check was on the plugins collection as a whole, therefore is the responsibility of the thing that _finds_ the plugins since it also knows what already exists. But again, hopefully a lot of this will start to fall into place once we chip away at more stages!

Reply:
Makes sense. Let's keep it like this and see where it goes! :+1: